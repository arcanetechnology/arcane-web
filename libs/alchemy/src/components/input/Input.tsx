/** @format */

import { VoidComponent, JSX, mergeProps, splitProps, Show } from 'solid-js';
import { InputVariant } from '../../types';
import { Label } from '../label/Label';

// so we dont have to create a base prop whenever we want to add new properties to base elements.
export type BaseInputProps = JSX.InputHTMLAttributes<HTMLInputElement>;

export const BaseInput: VoidComponent<BaseInputProps> = (props) => {
  const [local, others] = splitProps(props, ['id']);
  return <input id={`${local.id}-input`} {...others} />;
};

type OptionalInputProps = {
  variant: InputVariant;
  fullWidth: boolean;
  label: string;
};

type InputProps = {
  name: string;
} & Partial<OptionalInputProps> &
  BaseInputProps;

export const Input: VoidComponent<InputProps> = (props) => {
  const merged = mergeProps(
    { variant: 'standard', id: `arcane`, fullWidth: false },
    props
  );
  const [local, others] = splitProps(merged, [
    'id',
    'variant',
    'fullWidth',
    'children',
    'label',
    'name',
  ]);

  return (
    <>
      <Show when={local.label !== undefined}>
        <Label id={`${props.name}-${local.id}`} for={local.name}>
          {local.label}
        </Label>
      </Show>
      <BaseInput
        type="text"
        name={local.name}
        classList={{
          'radius-small': true,
          'w-full': local.fullWidth,
          'padding-2': true,
        }}
        id={`${props.name}-${local.id}`}
        {...others}
      />
    </>
  );
};
