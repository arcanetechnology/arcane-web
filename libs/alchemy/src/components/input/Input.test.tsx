/** @format */

import { Input } from './Input';
import { cleanup, render, screen } from 'solid-testing-library';

describe('alchemy input element', () => {
  afterEach(cleanup);

  it('renders input without full width if we do not full width', () => {
    render(() => <Input name="test-input" data-testid="test-input" />);
    const input = screen.getByTestId('test-input');
    expect(input).not.toHaveClass('w-full');
  });

  it('renders input with full width if we specify props with full width', () => {
    render(() => (
      <Input name="test-input" data-testid="test-input" fullWidth />
    ));
    const input = screen.getByTestId('test-input');
    expect(input).toHaveClass('w-full');
  });

  it('renders label element if label string is passed to the component', () => {
    const tree = render(() => (
      <Input name="test-input" data-testid="test" label="test-input" />
    )).baseElement;

    expect(tree).toMatchSnapshot();
  });
});
