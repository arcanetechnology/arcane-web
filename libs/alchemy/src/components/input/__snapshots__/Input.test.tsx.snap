// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`alchemy input element renders label element if label string is passed to the component 1`] = `
<body>
  <div>
    <label
      for="test-input"
      id="test-input-arcane-label"
    >
      test-input
    </label>
    <input
      class="radius-small padding-2"
      data-testid="test"
      id="test-input-arcane-input"
      name="test-input"
      type="text"
    />
  </div>
</body>
`;
